crates/worker/src/partition/state_machine/mod.rs

```diff
@@ -493,6 +482,43 @@ impl<Codec: RawEntryCodec> StateMachine<Codec> {
 
         }.await?;
 
+        // Handle duplicate requests from ingress RPCs. These are always deduplicated without further operations.
+        if let Some(ingress_sink @ ServiceInvocationResponseSink::Ingress { .. }) =
+            &service_invocation.response_sink
+        {
+            if let InvocationStatus::Scheduled(ScheduledInvocation {
+                metadata: PreFlightInvocationMetadata { response_sinks, .. },
+            })
+            | InvocationStatus::Inboxed(InboxedInvocation {
+                metadata: PreFlightInvocationMetadata { response_sinks, .. },
+                ..
+            })
+            | InvocationStatus::Invoked(InFlightInvocationMetadata {
+                response_sinks, ..
+            })
+            | InvocationStatus::Suspended {
+                metadata: InFlightInvocationMetadata { response_sinks, .. },
+                ..
+            } = &previous_invocation_status
+            {
+                if response_sinks.contains(ingress_sink) {
+                    return Ok(None);
+                }
+            }
+        }
```

Comment:
What if we handled non-idempotent invocations the same way we treat idempotent invocations? Like if an invocation is already present, then we update the `response_sinks` with then new `service_invocation.response_sink` and send a submit notification? This would make submitting invocations idempotently based on their invocation id independent of the idempotency key.

Now, we are already half-way there with this change.